python/ray/data/dataset.py

```diff
@@ -2411,7 +2411,24 @@ def to_mars(self) -> "mars.DataFrame":
         Returns:
             A MARS dataframe created from this dataset.
         """
-        raise NotImplementedError  # P1
+        import pandas as pd
+        import pyarrow as pa
+        from mars.dataframe.datasource.read_raydataset import DataFrameReadRayDataset
+        from mars.dataframe.utils import parse_index
+        from ray.data.impl.pandas_block import PandasBlockSchema
+
+        refs = self.to_pandas_refs()
+        # remove this when https://github.com/mars-project/mars/issues/2945 got fixed
+        if isinstance(self.schema(), PandasBlockSchema):
```

Comment:
Could you cache the schema fetch and reuse it throughout the rest of this function?
```suggestion
        schema = self.schema()
        if isinstance(schema, PandasBlockSchema):
```

Reply:
I'll fix it